File: library/src/main/java/com/opensource/svgaplayer/SVGAVideoSpriteEntity.java
package com.opensource.svgaplayer;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;
import java.util.ArrayList;

/**
 * Created by cuiminghui on 2016/10/17.
 */
class SVGAVideoSpriteEntity {

    String imageKey;
    ArrayList<SVGAVideoSpriteFrameEntity> frames;

    public SVGAVideoSpriteEntity() {
    }

    SVGAVideoSpriteEntity(JSONObject obj) throws JSONException {
        imageKey = obj.getString("imageKey");
        frames = new ArrayList<>();
        JSONArray jsonFrames = obj.getJSONArray("frames");
        for (int i = 0; i < jsonFrames.length(); i++) {
            JSONObject frameObject = jsonFrames.getJSONObject(i);
            SVGAVideoSpriteFrameEntity frameItem = new SVGAVideoSpriteFrameEntity(frameObject);
            if (frameItem.shapes.length == 1 && frameItem.shapes[0].isKeep() && frames.size() > 0) {
                frameItem.shapes = frames.get(frames.size() - 1).shapes;
            }
            frames.add(frameItem);
        }
    }

}

File: library/src/main/java/com/opensource/svgaplayer/SVGAVideoSpriteEntity.kt
package com.opensource.svgaplayer

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.io.Serializable
import java.util.ArrayList

/**
 * Created by cuiminghui on 2016/10/17.
 */
class SVGAVideoSpriteEntity {

    val imageKey: String?

    val frames: List<SVGAVideoSpriteFrameEntity>

    constructor(obj: JSONObject) {
        this.imageKey = obj.optString("imageKey")
        val mutableFrames: MutableList<SVGAVideoSpriteFrameEntity> = mutableListOf()
        obj.optJSONArray("frames")?.let {
            for (i in 0..it.length() - 1) {
                it.optJSONObject(i)?.let {
                    val frameItem = SVGAVideoSpriteFrameEntity(it)
                    if (frameItem.shapes.size > 0) {
                        frameItem.shapes.first()?.let {
                            if (it.isKeep && mutableFrames.size > 0) {
                                frameItem.shapes = mutableFrames.last().shapes
                            }
                        }
                    }
                    mutableFrames.add(frameItem)
                }
            }
        }
        frames = mutableFrames.toList()
    }

    constructor(obj: ComOpensourceSvgaVideo.SpriteEntity) {
        this.imageKey = obj.imageKey
        var lastFrame: SVGAVideoSpriteFrameEntity? = null
        frames = obj.framesList.map {
            val frameItem = SVGAVideoSpriteFrameEntity(it)
            if (frameItem.shapes.size > 0) {
                frameItem.shapes.first()?.let {
                    if (it.isKeep) {
                        lastFrame?.let {
                            frameItem.shapes = it.shapes
                        }
                    }
                }
            }
            lastFrame = frameItem
            return@map frameItem
        }

    }

}
